processing/src/main/java/io/druid/query/PrioritizedExecutorService.java

```diff
@@ -185,89 +190,215 @@ public int getQueueSize()
-    {
-      return new PrioritizedListenableFutureTask<>(task, priority);
-    }
+  private final ListenableFutureTask<V> delegate;
+  private final int priority;
 
-    private final ListenableFutureTask<V> delegate;
-    private final int priority;
+  PrioritizedListenableFutureTask(ListenableFutureTask<V> delegate, int priority)
+  {
+    this.delegate = delegate;
+    this.priority = priority;
+  }
 
-    PrioritizedListenableFutureTask(ListenableFutureTask<V> delegate, int priority)
-    {
-      this.delegate = delegate;
-      this.priority = priority;
-    }
+  @Override
+  public void run()
+  {
+    delegate.run();
+  }
 
-    @Override
-    public void run()
-    {
-      delegate.run();
-    }
+  @Override
+  public boolean cancel(boolean mayInterruptIfRunning)
+  {
+    return delegate.cancel(mayInterruptIfRunning);
+  }
 
-    @Override
-    public boolean cancel(boolean mayInterruptIfRunning)
-    {
-      return delegate.cancel(mayInterruptIfRunning);
-    }
+  @Override
+  public boolean isCancelled()
+  {
+    return delegate.isCancelled();
+  }
 
-    @Override
-    public boolean isCancelled()
-    {
-      return delegate.isCancelled();
-    }
+  @Override
+  public boolean isDone()
+  {
+    return delegate.isDone();
+  }
 
-    @Override
-    public boolean isDone()
-    {
-      return delegate.isDone();
-    }
+  @Override
+  public V get() throws InterruptedException, ExecutionException
+  {
+    return delegate.get();
+  }
 
-    @Override
-    public V get() throws InterruptedException, ExecutionException
-    {
-      return delegate.get();
-    }
+  @Override
+  public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
+  {
+    return delegate.get(timeout, unit);
+  }
 
-    @Override
-    public V get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException
-    {
-      return delegate.get(timeout, unit);
-    }
+  @Override
+  public void addListener(Runnable listener, Executor executor)
+  {
+    delegate.addListener(listener, executor);
+  }
 
-    @Override
-    public void addListener(Runnable listener, Executor executor)
-    {
-      delegate.addListener(listener, executor);
-    }
+  @Override
+  public int getPriority()
+  {
+    return priority;
+  }
+
+  @Override
+  public int compareTo(PrioritizedListenableFutureTask otherTask)
+  {
+    return Ints.compare(otherTask.getPriority(), getPriority());
+  }
+}
 
+class PrioritizedTaskBlockingQueue extends PriorityBlockingQueue<Runnable>
+{
+  private static final Comparator<Runnable> PRIORITY_AND_ENUMERATION_COMPARABLE = new Ordering<Runnable>()
+  {
     @Override
-    public int getPriority()
+    public int compare(@Nullable Runnable r1, @Nullable Runnable r2)
     {
-      return priority;
+      int compare = 0;
+      if (r1 instanceof Prioritized && r2 instanceof Prioritized) {
```

Comment:
I modified it a bit based on @xvrl 's suggestions. The `PrioritizedListenableFutureTask` uses a `ListenableFutureTask` which wraps the callable vs runnable of the underlying "thing which needs executing". I refactored how the priority is extracted just a bit, hopefully it looks cleaner now.